### DISCUSSION TITLE
Formatters/Linters setup

### DISCUSSION THREAD

**Discussion by kobzar** [LazyVim/LazyVim#751] [Q&A]
### Did you check the docs?

- [X] I have read all the LazyVim docs

### Is your feature request related to a problem? Please describe.

Please help me understand how to setup linters and formatters for diff langs
for example for python I am use "isort, black, flake8"
I don't use options format on save, so I want to format buffer by key combinatio
like ```map("<leader>ff", format fille)```

Thanks

### Describe the solution you'd like

Add examples on homepage

### Describe alternatives you've considered

This is default in lunar vim

**Comment by dpetka2001** [comment]
Did you read the docs? All you need is [mason](https://www.lazyvim.org/plugins/lsp#masonnvim) and [null-ls](https://www.lazyvim.org/plugins/lsp#null-lsnvim). As for format on save there is keybinding to toggle it on/off depending on your needs.

**Comment by kobzar** [comment]
I fiund two topics in discusions here but dont find working example for any of langs. Only teoretical ideas

**Reply by dpetka2001** [reply]
I'm pretty sure there are examples of configs as well. There is also configs in the [docs](https://www.lazyvim.org/plugins/extras/formatting.prettier). Read them to understand how it's done. This is only for `prettier` example which is just a formatter used by many languages. There's also example for [Typescript](https://www.lazyvim.org/plugins/extras/lang.typescript) which you can use to set up other languages as well.

**Comment by kobzar** [comment]
I dont need typescript or prettier.
I want use flake8 + black + isort for python files. And i whant to make format document by keybind.
```lua
require("lazy").setup({
  spec = {
    { "LazyVim/LazyVim", import = "lazyvim.plugins" },
    { import = "lazyvim.plugins.extras.lang.typescript" },
    { import = "lazyvim.plugins.extras.formatting.prettier" },
    { import = "plugins" },
  },
})
```
But black has a lot personal settings.  And laso i want use black only for python files. For rust i want use rust-analyzer + rustfmt 
Sorry, but its really hard to understand and make it without examples.

```
  { import = "lazyvim.plugins.extras.lang.typescript" },
    { import = "lazyvim.plugins.extras.formatting.prettier" },
```
Where and how i can get full list of params for this strings?

**Reply by dpetka2001** [reply]
I didn't say you need Typescript or prettier. I just mentioned that you can use them as examples to derive how you can set up the languages you want. For example this is how I have Python set up:
```lua
return {
  {
    "williamboman/mason.nvim",
    opts = function(_, opts)
      vim.list_extend(opts.ensure_installed, {
        "pyright",
        "black",
        "ruff-lsp",
        "ruff",
      })
    end,
  },

  -- Setup null-ls with `black`
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      local nls = require("null-ls")
      opts.sources = vim.list_extend(opts.sources, {
        nls.builtins.formatting.black,
        nls.builtins.formatting.ruff,
        -- nls.builtins.diagnostics.ruff,
      })
    end,
  },
}
```
You can add `flake8` and `isort` to mason and then set them up in null-ls.

**Reply by mohit2152sharma** [reply]
@dpetka2001 where do you place this file, I mean which directory?

**Reply by dpetka2001** [reply]
In some file or directory (depending on how you want to structure your configuration) in `~/.config/nvim/lua/plugins` directory.
